src/adapters/bitbucket.js

```diff
@@ -0,0 +1,255 @@
+      })
+    }
+    else { // falls back
+      super.selectFile(path)
+    }
+  }
+
+  // @override
+  downloadFile(path, fileName) {
+    const link = document.createElement('a')
+    link.setAttribute('href', path.replace(/\/src\//, '/raw/'))
+    link.setAttribute('download', fileName)
+    link.click()
+  }
+
+  // @override
+  loadCodeTree(opts, cb) {
+    opts.path = opts.node.path
+    this._loadCodeTree(opts, (item) => {
+      if (!item.type) {
+        item.type = 'blob'
+      }
+    }, cb)
+  }
+
+  // @override
+  _getTree(path, opts, cb) {
+    this._get(`/src/${opts.repo.branch}/${path}`, opts, (err, res) => {
+      if (err) return cb(err)
+      const directories = res.directories.map((dir) => ({path: dir, type: 'tree'}))
+      res.files.forEach((file) => {
+        if (file.path.startsWith(res.path)) {
+          file.path = file.path.substring(res.path.length)
+        }
+      })
+      const tree = res.files.concat(directories)
+      cb(null, tree)
+    })
+  }
+
+  // @override
+  _getSubmodules(tree, opts, cb) {
+    if (opts.repo.submodules) {
+      return this._getSubmodulesInCurrentPath(tree, opts, cb)
+    }
+
+    const item = tree.filter((item) => /^\.gitmodules$/i.test(item.path))[0]
+    if (!item) return cb()
+
+    this._get(`/src/${opts.encodedBranch}/${item.path}`, opts, (err, res) => {
+      if (err) return cb(err)
+      // Memoize submodules so that they will be inserted into the tree later.
+      opts.repo.submodules = parseGitmodules(res.data)
+      this._getSubmodulesInCurrentPath(tree, opts, cb)
+    })
+  }
+
+  _getSubmodulesInCurrentPath(tree, opts, cb) {
+    const currentPath = opts.path
+    const isInCurrentPath = currentPath
+      ? (path) => path.startsWith(`${currentPath}/`)
+      : (path) => path.indexOf('/') === -1
+
+    const submodules = opts.repo.submodules
+    const submodulesInCurrentPath = {}
+    Object.keys(submodules).filter(isInCurrentPath).forEach((key) => {
+      submodulesInCurrentPath[key] = submodules[key]
+    })
+
+    // Insert submodules in current path into the tree because submodules can not
+    // be retrieved with Bitbucket API but can only by reading .gitmodules.
+    Object.keys(submodulesInCurrentPath).forEach((path) => {
+      if (currentPath) {
+        // `currentPath` is prefixed to `path`, so delete it.
+        path = path.substring(currentPath.length + 1)
+      }
+      tree.push({path: path, type: 'commit'})
+    })
+    cb(null, submodulesInCurrentPath)
+  }
+
+  _get(path, opts, cb) {
+    const host = location.protocol + '//' + 'api.bitbucket.org/1.0'
+    const url = `${host}/repositories/${opts.repo.username}/${opts.repo.reponame}${path || ''}`
+    const cfg  = { url, method: 'GET', cache: false }
+
+    if (opts.token) {
+      // Bitbucket App passwords can be used only for Basic Authentication.
+      // Get username of logged-in user.
+      const currentUser = JSON.parse($('body').attr('data-current-user'))
```

Comment:
Maybe it's more reliable if we allow user to specify both username & password. We can use a single field a separate by `:`. Just change the verbiage in the case of BitBucket so that it's clear.

In other words, if the token has `:`, assume users enter both and use this username. If users don't explicitly specify username, retrieve it from the page (like currently). If can't retrieve (DOM changes), show error and ask user to enter both username & password, separated by `:`.